Control-click on YouTube video does not show video controls

lilyspiniolas opened the pull request:
#### 5336d0544d6b8a9d3db284ad17b7df75584d0141
<pre>
Control-click on YouTube video does not show video controls
<a href="https://bugs.webkit.org/show_bug.cgi?id=301730">https://bugs.webkit.org/show_bug.cgi?id=301730</a>
<a href="https://rdar.apple.com/162276991">rdar://162276991</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

After fractional coordinates were enabled in 298383@main,
`ContextMenuController::maybeCreateContextMenu` converted the absolute location
of the mouse from a DoublePoint to a LayoutPoint. Previously, it was an IntPoint
which was converted to a LayoutPoint.

This difference is responsible for causing right-clicking or control-clicking
on a YouTube video to fail to show video controls because the hit test began
to return the element used for YouTube&apos;s own context menu rather than hitting
the video.

We fix this by effectively restoring the old behavior; the DoublePoint absolute
location is now floored before it is converted to a LayoutPoint.

While this fix results in a loss of precision, it is arguably more correct to
floor the point because we currently do not report fractional coordinates for
contextmenu events, and the hit test now reflects this by using the exact same
point that we report for the event coordinates.

`eventSender.contextClick()` now respects fractional coordinates for the mouse
position instead of flooring the location.

Test: fast/events/contextmenu-items-match-node-receiving-contextmenu-event.html

* LayoutTests/fast/events/contextmenu-items-match-node-receiving-contextmenu-event-expected.txt: Added.
* LayoutTests/fast/events/contextmenu-items-match-node-receiving-contextmenu-event.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::maybeCreateContextMenu):
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toDoublePoint):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageCopyContextMenuAtPointInWindow):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::contextMenuAtPointInWindow):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/302489@main">https://commits.webkit.org/302489@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdac931f283651b5f664e204228c988c0612cf05

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40073 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136613 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80624 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5ccf6e9c-6bdf-408e-ace7-e05ec09c6a62) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131108 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1426 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1370 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98410 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66313 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/394efd29-8026-4a92-8f30-85a34eb90a39) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1109 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115762 "Found 1 new API test failure: TestWebKitAPI.SiteIsolation.NavigationAfterWindowOpen (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79060 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0486b2f1-18a4-46ec-bdf5-d63ce7d49216) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1035 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33881 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79892 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109487 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34375 "Found 1 new test failure: http/tests/site-isolation/window-open-with-name-cross-site.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139086 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1284 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1238 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106941 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112098 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106779 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1060 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30619 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53889 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20179 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1356 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64712 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1180 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1218 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1279 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on LayoutTests/fast/events/contextmenu-items-match-node-receiving-contextmenu-event.html:
```diff
@@ -0,0 +1,81 @@
+</head>
+<body>
+    <p>The node which receives a 'contextmenu' event should be the node hit-tested when determining which items to show.</p>
+    <video id="video" src="resources/test.mp4"></video>
```
nit, mainly because 188 KB is not tiny, I feel like you could get away with any of the other pre-existing `.mp4` test resources? Instead of checking in a new file?

```suggestion
    <video id="video" src="../../media/resources/white.mp4"></video>
```

lilyspiniolas replied:
I know this probably is not a super likely scenario, but I decided against this originally out of concern that, if the media tests are ever updated to not use that video and it ends up being removed, it won't be obvious that this test was relying on it as well because the test would continue to pass.

lilyspiniolas replied:
I think I'll just add some JS to fail the test if the file can't be loaded, just to be safe. That way we can reuse an existing file without any false passes.

lilyspiniolas committed:
Control-click on YouTube video does not show video controls
https://bugs.webkit.org/show_bug.cgi?id=301730
rdar://162276991

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

After fractional coordinates were enabled in 298383@main,
`ContextMenuController::maybeCreateContextMenu` converted the absolute location
of the mouse from a DoublePoint to a LayoutPoint. Previously, it was an IntPoint
which was converted to a LayoutPoint.

This difference is responsible for causing right-clicking or control-clicking
on a YouTube video to fail to show video controls because the hit test began
to return the element used for YouTube's own context menu rather than hitting
the video.

We fix this by effectively restoring the old behavior; the DoublePoint absolute
location is now floored before it is converted to a LayoutPoint.

While this fix results in a loss of precision, it is arguably more correct to
floor the point because we currently do not report fractional coordinates for
contextmenu events, and the hit test now reflects this by using the exact same
point that we report for the event coordinates.

`eventSender.contextClick()` now respects fractional coordinates for the mouse
position instead of flooring the location.

Test: fast/events/contextmenu-items-match-node-receiving-contextmenu-event.html

* LayoutTests/fast/events/contextmenu-items-match-node-receiving-contextmenu-event-expected.txt: Added.
* LayoutTests/fast/events/contextmenu-items-match-node-receiving-contextmenu-event.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::maybeCreateContextMenu):
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toDoublePoint):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageCopyContextMenuAtPointInWindow):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::contextMenuAtPointInWindow):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/302489@main
